scipy/_lib/_array_api.py

```diff
@@ -0,0 +1,103 @@
+"""Utility functions to use Python Array API compatible libraries.
+
+For the context about the Array API see:
+https://data-apis.org/array-api/latest/purpose_and_scope.html
+
+The SciPy use case of the Array API is described on the following page:
+https://data-apis.org/array-api/latest/use_cases.html#use-case-scipy
+"""
+import os
+
+import numpy as np
+import numpy.array_api
+# probably want to vendor it (submodule)
+import array_api_compat
+import array_api_compat.numpy
+
+# SCIPY_ARRAY_API, array_api_dispatch is used by sklearn
+USE_ARRAY_API = os.environ.get("array_api_dispatch", False)
+
+__all__ = ['namespace_from_arrays', 'asarray', 'asarray_namespace']
+
+
+def namespace_from_arrays(*arrays, single_namespace=True):
+    # if we cannot get the namespace, np is used
+    namespaces = set()
+    for array in arrays:
+        try:
+            namespaces.add(array_api_compat.array_namespace(array))
```

Comment:
Yes I would need the function to not raise but return NumPy. This is because we want to accept things like Python lists or scalar.

Reply:
I think we can add a flag to `array_namespace` like `default=numpy` that makes it do this. That function's not part of the spec so we can adjust it however makes it most useful.

Reply:
That would help with our use case 😃

Quick question, should this really fail? 

```python
array_api_compat.array_namespace(np.array(1), numpy.array_api.asarray(1))
```
This is why I have the following to go around and convert `numpy.array_api` to `array_api_compat.numpy`
```python
  if numpy.array_api in namespaces:
      namespaces.remove(numpy.array_api)
      namespaces.add(array_api_compat.numpy)
```

Reply:
And in general any compat code that's useful to more than one library can go in the compat library (assuming it's not too complex and pure Python).

Reply:
ok I can do a PR for that if you want.

Reply:
I wouldn't try mixing numpy and numpy.array_api. numpy.array_api should only be used for testing purposes. It implements a strict version of the standard so you can use it to check that you are saying within the spec. It shouldn't be used for actual user code. 

The whole purpose of the compat library is to provide sufficient wrappers around numpy itself to make it array API compatible. Using numpy.array_api for user code was found to be too challenging because it uses a different array class from NumPy, which is not what most users want.

Reply:
mmm ok so when I test, I should consider that `numpy.array_api` is something different than `numpy` such as lets say `cupy`. Makes sense to me 👍 Thanks for the explanations Aaron 😃

Reply:
@asmeurer I opened https://github.com/data-apis/array-api-compat/pull/39